lib/valueflow.cpp

```diff
@@ -2468,67 +2468,160 @@ static bool valueFlowForward(Token * const               startToken,
     return true;
 }
 
+static const Variable * getLifetimeVariable(const Token * tok, ErrorPath& errorPath)
+{
+    const Variable * var = tok->variable();
+    if (!var)
+        return nullptr;
+    if (var->isReference() || var->isRValueReference()) {
+        for (const ValueFlow::Value& v:tok->values()) {
+            if (!v.isLifetimeValue())
+                continue;
+            if (v.tokvalue == tok)
+                continue;
+            errorPath.insert(errorPath.end(), v.errorPath.begin(), v.errorPath.end());
+            const Variable * var2 = getLifetimeVariable(v.tokvalue, errorPath);
+            if (var2)
+                return var2;
+        }
+        return nullptr;
+    }
+    return var;
+}
+
 static bool isNotLifetimeValue(const ValueFlow::Value& val)
 {
     return !val.isLifetimeValue();
 }
 
+static void valueFlowLifetimeFunction(Token * tok, TokenList *tokenlist, ErrorLogger *errorLogger, const Settings *settings);
+
 static void valueFlowForwardLifetime(Token * tok, TokenList *tokenlist, ErrorLogger *errorLogger, const Settings *settings)
 {
-    const Token * assignTok = tok->astParent();
-    // Assignment
-    if (!assignTok || (assignTok->str() != "=") || assignTok->astParent())
-        return;
-
-    // Lhs should be a variable
-    if (!assignTok->astOperand1() || !assignTok->astOperand1()->varId())
-        return;
-    const Variable *var = assignTok->astOperand1()->variable();
-    if (!var || (!var->isLocal() && !var->isGlobal() && !var->isArgument()))
+    const Token * parent = tok->astParent();
+    while(parent && parent->isArithmeticalOp())
+        parent = parent->astParent();
+    if(!parent)
         return;
+    // Assignment
+    if(parent->str() == "=" && !parent->astParent()) {
+        // Lhs should be a variable
+        if (!parent->astOperand1() || !parent->astOperand1()->varId())
```

Comment:
don't we check that `=` has both lhs and rhs? as I remember you added code for it.

Reply:
I am not against this check here. feel free to keep it if you prefer to have it..